### ISSUE TITLE
[fluxv2] Tag HelmRepository name in Catalog

### ISSUE THREAD

**Issue by hongkunyoo** [kubeapps/kubeapps#4231]
### Description:

I am planing to use only fluxv2 plugin.
I have two different Airflow chart in two different `HelmRepository`(let's say one for prod, one for dev).
Is it possible to show `HelmRepository` name which chart is coming from which?
Just like helm plugin that shows the name of the repo.

![스크린샷 2022-02-03 오후 10 35 24](https://user-images.githubusercontent.com/3633188/152354861-40f5dcd2-834b-41c4-a330-952172c8221f.png)

Is there any option that I missed to enable this feature?
Please guide me if so.

I really appreciate your effort!

**Comment by antgamdia** [comment]
Thanks for testing the Fluxv2 support out :) It's great to know there are people using it (even if it is still in a developer preview stage).

You have just run into this //TODO :stuck_out_tongue: https://github.com/kubeapps/kubeapps/blob/main/dashboard/src/components/Catalog/PackageCatalogItem.tsx#L30

The point is that we currently don't have any well-defined API for interacting with package repositories -yet- (follow the process in #3496 ), so getting the repository this package comes from is not possible yet.

However, let's ping @gfichtenholt  in case he has any idea to work around this pending item.

**Comment by gfichtenholt** [comment]
True, PackageRepository API is TBD at the moment.
However, I don't think there are plans to provide such a reference. Many months ago I asked the same question. As far as I remember, @absoludity told me that, in general, it is not possible, as there are some packaging systems that simply do not provide this info. And it was a generic API we were defining. Meaning, the lowest common denominator that works with all plug ins (helm, flux, carvel) - if some feature isn't supported by one of the plugins, it is out of scope for MVP. Something along these lines, my recollection is a bit hazy, this was a long time ago. If Michael remembers better, hopefully he can add some context to this.
Also, see comment right above https://github.com/kubeapps/kubeapps/blob/main/cmd/kubeapps-apis/proto/kubeappsapis/core/packages/v1alpha1/packages.proto#L713

**Comment by absoludity** [comment]
> Many months ago I asked the same question. As far as I remember, @absoludity told me that in general it is not possible, as there are some packaging systems that simply do not provide this info. 

Yes, it was the Carvel support, and last time we asked about this, Carvel didn't (by design, I believe) have a link from a package available for install ([package meta or package](https://carvel.dev/kapp-controller/docs/v0.32.0/packaging/#package-metadata)) to the repository (and still doesn't). That said, this has been requested not just by us, but others as well. You can read the details on https://github.com/vmware-tanzu/carvel-kapp-controller/issues/124 , one use-case there is exactly @hongkunyoo 's case (having a dev and prod package from two different repos). So it may be that they end up providing this (even if it's by an annotation).

So it seems to be something that people have a use-case for, across different packaging systems and would be excellent to support in Kubeapps. We currently have a work-around for this with Helm in the frontend (via the identifier, which has the repo encoded as mentioned in the link to which Greg referred). For flux, also as Greg mentioned, it's supported out of the box (in that the available package corresponds to the [HelmChart CR](https://fluxcd.io/docs/components/source/helmcharts/) which links to the source repo or other for the chart).

Given that, I have no issue with us considering adding a field to the `AvailablePackage` message, such as `Source` (using the more general flux usage, [since the source can be a helm repo, a git repo or bucket etc.](https://fluxcd.io/docs/components/source/helmcharts/), given that we can already use it for both Helm and Flux, and it seems eventually Carvel too (no issue for it to be blank/unused for the carvel plugin for now). We could do this without depending on the later app repo api design, but need to keep it in mind so that the format of the field would not be a backwards step towards a future goal, IMO.

**Comment by gfichtenholt** [comment]
Agree 99% with what Michael said. The only thing I would change is not call it `Source` but call it `PackageRepositoryReference`. That way it is consistent with what I plan do for PackageRepository API. Another option that works for me is to have Source and have PackageRepositoryReference inherit from Source (i.e. "is a"-relationship). In other words, I am attached to the word "PackageRepositoryReference", since its been reviewed and we start using some other term now, things will be difficult in the future.  Caveat: Don't know of go/protobuf supports inheritance. If not may be we can hack around with composition instead, i.e. "has a"-relationship

**Comment by absoludity** [comment]
Great, I'm fine with `PackageRepositoryReference`, I only mentioned source because I was reading the flux docs and realised that they use that because the source may not be a repository (ie. could just be a bucket, or git).

> Regarding "We currently have a work-around for this with Helm in the frontend (via the identifier, which has the repo encoded as mentioned in the link to which Greg referred)". Just FYI, flux happens to use the same encoding `repoName/packageName`, so feel free to apply the same hack in the front end to flux smile

Sweet.

**Comment by gfichtenholt** [comment]
> Great, I'm fine with `PackageRepositoryReference`, I only mentioned source because I was reading the flux docs and realised that they use that because the source may not be a repository (ie. could just be a bucket, or git).
> 
> > Regarding "We currently have a work-around for this with Helm in the frontend (via the identifier, which has the repo encoded as mentioned in the link to which Greg referred)". Just FYI, flux happens to use the same encoding `repoName/packageName`, so feel free to apply the same hack in the front end to flux smile
> 
> Sweet.

FWIW, git in flux  is still a _repository_ https://fluxcd.io/docs/components/source/gitrepositories/
S3 buckets are not, but if you squint hard enough, you probably can get away with calling a "Repository"

**Comment by vrabbi** [comment]
FYI, kapp controller 0.33.1 added a reference to the package repo via an annotation:
"pkgr: add annotation showing source pkgr to created resources."
This should allow the same behavior for carvel as the other plugins as long as people are using 0.33.1 or above.

**Comment by antgamdia** [comment]
Thank you! Will look into it!
